src/admin/upload_index.rs

```diff
@@ -0,0 +1,50 @@
+use crate::admin::dialoguer;
+use cargo_registry_index::{Repository, RepositoryConfig};
+use reqwest::blocking::Client;
+
+use crate::config;
+
+#[derive(clap::Parser, Debug)]
+#[clap(
+    name = "upload-index",
+    about = "Upload index from git to S3 (http-based index)"
+)]
+pub struct Opts {
+    /// Incremental commit. Any changed files made after this commit will be uploaded.
+    incremental_commit: Option<String>,
+}
+
+pub fn run(opts: Opts) -> anyhow::Result<()> {
+    let config = config::Base::from_environment();
+    let uploader = config.uploader();
+    let client = Client::new();
+
+    println!("fetching git repo");
+    let config = RepositoryConfig::from_environment();
+    let repo = Repository::open(&config)?;
+    repo.reset_head()?;
+    println!("HEAD is at {}", repo.head_oid()?);
+
+    let files = repo.get_files_modified_since(opts.incremental_commit.as_deref())?;
+    println!("found {} files to upload", files.len());
+    if !dialoguer::confirm("continue with upload?") {
+        return Ok(());
+    }
+
+    for file in files {
+        let crate_name = file.file_name().unwrap().to_str().unwrap();
+        let path = repo.index_file(crate_name);
+        if !path.exists() {
+            println!("skipping file `{}`", crate_name);
+            continue;
+        }
+        let contents = std::fs::read_to_string(&path)?;
+        uploader.upload_index(&client, crate_name, contents)?;
+    }
```

Comment:
do you have any estimates how long this operation might take? would it make sense to add some sort of progress logging to the loop?

Reply:
It needs to upload ~83000 small files to S3. I would guess around an hour, but of course it depends on the internet connection.

I'll add progress logging.